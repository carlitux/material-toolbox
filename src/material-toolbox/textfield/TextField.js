// @flow
import * as React from 'react';
import * as util from '@material/ripple/util';
import cx from 'classnames';

import MDCTextFieldFoundation from '@material/textfield/foundation';
import MDCRippleFoundation from '@material/ripple/foundation';

import FloatingLabel from '../floating-label';
import LineRipple from '../line-ripple';
import NotchedOutline from '../notched-outline';
import HelperText from './HelperText';
import { createAdapter } from '../ripple';

type Props = {
  label: string,
  cssOnly: boolean,
  fullWidth: boolean,
  multiline: boolean,
  disabled: boolean,
  dense: boolean,
  outlined: boolean,
  box: boolean,
  helperText?: string,
  persistentHelperText: boolean,
  validationHelperText: boolean,
  value?: string,
  defaultValue?: string,
  invalid: boolean,
  className?: string,
  'aria-controls'?: string,
  onClick?: (evt: Event) => void,
  onKeyDown?: (evt: Event) => void,
  children?: React.Element<any>,
  icon?: 'trailing' | 'leading',
};

type State = {
  classNames: { [string]: string },
  rippleStyle: { [string]: any },
  rippleClassName: { [string]: boolean },
};

// TODO: Move ref to the new way of references
export default class TextField extends React.Component<Props, State> {
  static defaultProps = {
    cssOnly: false,
    fullWidth: false,
    multiline: false,
    disabled: false,
    dense: false,
    invalid: false,
    outlined: false,
    persistentHelperText: false,
    validationHelperText: false,
    box: false,
  };

  state = {
    rippleClassName: {},
    rippleStyle: {},
    classNames: {},
  };

  componentDidMount() {
    if (!this.props.cssOnly) {
      this.createMDCTextField_();
      this.createMDCRipple_(this.props);
    }
  }

  componentDidUpdate(prevProps: Props) {
    const didChange = [
      'multiline',
      'helperText',
      'cssOnly',
      'outlined',
      'box',
      'aria-controls',
    ].some(key => this.props[key] !== prevProps[key]);

    if (didChange) {
      this.removeMDCRipple_();
      this.removeMDCTextField_();
      if (!this.props.cssOnly) {
        this.createMDCTextField_();
        this.createMDCRipple_(this.props);
      }
    }

    if (this.mdcTextField_ && this.props.invalid !== prevProps.invalid) {
      this.mdcTextField_.setValid(!this.props.invalid);
    }
  }

  componentWillUnmount() {
    this.removeMDCRipple_();
    this.removeMDCTextField_();
  }

  getInputAdapterMethods_() {
    return {
      registerInputInteractionHandler: (evtType: any, handler: () => void) =>
        this.input_ && this.input_.addEventListener(evtType, handler),
      deregisterInputInteractionHandler: (evtType: any, handler: () => void) =>
        this.input_ && this.input_.removeEventListener(evtType, handler),
      getNativeInput: () => this.input_,
    };
  }

  getLabelAdapterMethods_() {
    return {
      shakeLabel: (shouldShake: boolean) =>
        this.label_ && this.label_.shake(shouldShake),
      floatLabel: (shouldFloat: boolean) =>
        this.label_ && this.label_.float(shouldFloat),
      hasLabel: () => !!this.label_,
      getLabelWidth: () => this.label_ && this.label_.getWidth(),
    };
  }

  getLineRippleAdapterMethods_() {
    return {
      activateLineRipple: () => {
        if (this.lineRipple_) {
          this.lineRipple_.activate();
        }
      },
      deactivateLineRipple: () => {
        if (this.lineRipple_) {
          this.lineRipple_.deactivate();
        }
      },
      setLineRippleTransformOrigin: (normalizedX: number) => {
        if (this.lineRipple_) {
          this.lineRipple_.setRippleCenter(normalizedX);
        }
      },
    };
  }

  getOutlineAdapterMethods_() {
    return {
      notchOutline: (labelWidth: number, isRtl: boolean) =>
        this.notchedOutline_ && this.notchedOutline_.notch(labelWidth, isRtl),
      closeOutline: () =>
        this.notchedOutline_ && this.notchedOutline_.closeNotch(),
      hasOutline: () => !!this.notchedOutline_,
    };
  }

  getFoundationMap_() {
    return {
      helperText: {
        showToScreenReader: () =>
          this.helperText_ && this.helperText_.showToScreenReader(),
        setValidity: (inputIsValid: boolean) =>
          this.helperText_ && this.helperText_.setValidity(inputIsValid),
      },
    };
  }

  createMDCTextField_() {
    this.mdcTextField_ = new MDCTextFieldFoundation(
      {
        addClass: className =>
          this.setState(state => ({
            classNames: { ...state.classNames, [className]: true },
          })),
        removeClass: className =>
          this.setState(state => ({
            classNames: { ...state.classNames, [className]: undefined },
          })),
        hasClass: className => this.state.classNames[className],
        registerTextFieldInteractionHandler: (evtType, handler) => {
          this.handlers[evtType] = handler;
        },
        deregisterTextFieldInteractionHandler: evtType =>
          delete this.handlers[evtType],
        registerValidationAttributeChangeHandler: handler => {
          const observer = new MutationObserver(handler);
          if (this.input_) {
            const targetNode = this.input_;
            const config = { attributes: true };
            observer.observe(targetNode, config);
          }
          return observer;
        },
        deregisterValidationAttributeChangeHandler: observer =>
          observer.disconnect(),
        isFocused: () => document.activeElement === this.input_,
        isRtl: () =>
          window.getComputedStyle(this.root_).getPropertyValue('direction') ===
          'rtl',
        // Extra
        ...this.getInputAdapterMethods_(),
        ...this.getLabelAdapterMethods_(),
        ...this.getLineRippleAdapterMethods_(),
        ...this.getOutlineAdapterMethods_(),
      },
      this.getFoundationMap_(),
    );
    this.mdcTextField_.init();
  }

  createMDCRipple_(props: Props) {
    if (props.box) {
      const MATCHES = util.getMatchesProperty(HTMLElement.prototype);

      this.rippleAdapter_ = Object.assign(createAdapter(this, this.root_, {}), {
        // $FlowFixMe
        isSurfaceActive: () => this.input_[MATCHES](':active'),
        registerInteractionHandler: (type, handler) =>
          this.input_ && this.input_.addEventListener(type, handler),
        deregisterInteractionHandler: (type, handler) =>
          this.input_ && this.input_.removeEventListener(type, handler),
      });

      this.mdcRipple_ = new MDCRippleFoundation(this.rippleAdapter_);
      this.mdcRipple_.init();
    }
  }

  removeMDCRipple_() {
    if (this.rippleAdapter_) {
      this.rippleAdapter_.componentRemoved();
    }

    if (this.mdcRipple_) {
      this.mdcRipple_.destroy();
      delete this.mdcRipple_;
      delete this.rippleAdapter_;
    }
  }

  removeMDCTextField_() {
    if (this.mdcTextField_) {
      this.mdcTextField_.destroy();
      delete this.mdcTextField_;
    }
  }

  handleClick = (evt: Event) => {
    if (this.handlers.click) {
      this.handlers.click(evt);
    }
    if (this.props.onClick) {
      this.props.onClick(evt);
    }
  };

  handleKeyDown = (evt: Event) => {
    if (this.handlers.keydown) {
      this.handlers.keydown(evt);
    }
    if (this.props.onKeyDown) {
      this.props.onKeyDown(evt);
    }
  };

  mdcRipple_: ?MDCRippleFoundation;
  mdcTextField_: ?MDCTextFieldFoundation;
  handlers: { [string]: (evt: Event) => void } = {};
  rippleAdapter_: ?any;

  // Refs
  label_: ?FloatingLabel;
  lineRipple_: ?LineRipple;
  helperText_: ?HelperText;
  notchedOutline_: ?NotchedOutline;
  input_: ?HTMLInputElement | ?HTMLTextAreaElement;
  root_: ?HTMLLabelElement | ?HTMLDivElement;

  render() {
    const {
      multiline,
      fullWidth,
      className,
      cssOnly,
      value,
      label,
      disabled,
      dense,
      invalid,
      helperText,
      persistentHelperText,
      validationHelperText,
      outlined,
      box,
      children,
      icon,
      'aria-controls': ariaControls,
      ...rest
    } = this.props;
    const Component = cssOnly ? 'div' : 'label';
    const Input = multiline ? 'textarea' : 'input';

    const rootClassName = cx(
      'mdc-text-field',
      className,
      this.state.classNames,
      this.state.rippleClassName,
      {
        'mdc-text-field--upgraded': value,
        'mdc-text-field--box': box,
        'mdc-text-field--outlined': outlined,
        'mdc-text-field--fullwidth': fullWidth,
        'mdc-text-field--textarea': multiline,
        'mdc-text-field--disabled': disabled,
        'mdc-text-field--dense': dense,
        [`mdc-text-field--with-${icon || ''}-icon`]: icon,
      },
    );

    // Input props
    const inputProps = {
      ...rest,
      disabled,
      value,
      // if input and fullFill or cssOnly enable placeholder
      placeholder: cssOnly || (fullWidth && !multiline) ? label : null,
      className: 'mdc-text-field__input',
      'aria-controls': ariaControls,
      ref: element => {
        this.input_ = element;
      },
    };

    if ((box || outlined) && fullWidth) {
      // eslint-disable-next-line
      console.error('You can\'t use box or outlined with fullWidth');
    }

    return (
      <React.Fragment>
        <Component
          style={this.state.rippleStyle}
          ref={element => {
            this.root_ = element;
          }}
          onClick={this.handleClick}
          onKeyDown={this.handleKeyDown}
          className={rootClassName}>
          {children &&
            icon === 'leading' && (
              <children.type
                {...children.props}
                disabled={disabled}
                cssOnly={cssOnly}
              />
            )}
          <Input {...inputProps} />
          {(!fullWidth || multiline) && (
            <FloatingLabel
              ref={element => {
                this.label_ = element;
              }}
              label={label}
              cssOnly={cssOnly}
              preFilled={!!this.props.defaultValue}
            />
          )}
          {children &&
            icon === 'trailing' && (
              <children.type
                {...children.props}
                disabled={disabled}
                cssOnly={cssOnly}
              />
            )}
          {!outlined &&
            !multiline && (
              <LineRipple
                ref={element => {
                  this.lineRipple_ = element;
                }}
                cssOnly={cssOnly}
              />
            )}
          {outlined && (
            <NotchedOutline
              ref={element => {
                this.notchedOutline_ = element;
              }}
              cssOnly={cssOnly}
            />
          )}
        </Component>
        {helperText && (
          <HelperText
            ref={element => {
              this.helperText_ = element;
            }}
            id={ariaControls}
            text={helperText}
            cssOnly={cssOnly}
            persistent={persistentHelperText}
            validation={validationHelperText}
          />
        )}
      </React.Fragment>
    );
  }
}
