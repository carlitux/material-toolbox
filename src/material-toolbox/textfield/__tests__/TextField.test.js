// @flow
import React from 'react';
import { shallow } from 'enzyme';
import TextField from '../TextField';
import TextFieldIcon from '../Icon';

global.MutationObserver = function MutationObserver() {};

describe('component::TextField', () => {
  it('Should match snapshoot', () => {
    const component = shallow(<TextField label="test" />);
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot cssOnly', () => {
    const component = shallow(<TextField label="test" cssOnly />);
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot fullWidth', () => {
    const component = shallow(<TextField label="test" fullWidth />);
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot multiline', () => {
    const component = shallow(<TextField label="test" multiline />);
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot disabled', () => {
    const component = shallow(<TextField label="test" disabled />);
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot dense', () => {
    const component = shallow(<TextField label="test" dense />);
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot outlined', () => {
    const component = shallow(<TextField label="test" outlined />);
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot box', () => {
    const component = shallow(<TextField label="test" box />);
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot value', () => {
    const component = shallow(<TextField label="test" value="test" />);
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot invalid', () => {
    const component = shallow(<TextField label="test" invalid />);
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot aria-controls', () => {
    const component = shallow(<TextField label="test" aria-controls="test" />);
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot helperText', () => {
    const component = shallow(
      <TextField
        label="test"
        helperText="this is a helper text"
        aria-controls="test"
      />,
    );
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot persistentHelperText', () => {
    const component = shallow(
      <TextField
        label="test"
        helperText="this is a helper text"
        persistentHelperText
        aria-controls="test"
      />,
    );
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot validationHelperText', () => {
    const component = shallow(
      <TextField
        label="test"
        helperText="this is a helper text"
        validationHelperText
        invalid
      />,
    );
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot icon leading', () => {
    const component = shallow(
      <TextField label="test" icon="leading" box>
        <TextFieldIcon className="material-icons">delete</TextFieldIcon>
      </TextField>,
    );
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot icon trailing', () => {
    const component = shallow(
      <TextField label="test" icon="trailing" outlined>
        <TextFieldIcon className="material-icons">delete</TextFieldIcon>
      </TextField>,
    );
    expect(component).toMatchSnapshot();
  });
});
