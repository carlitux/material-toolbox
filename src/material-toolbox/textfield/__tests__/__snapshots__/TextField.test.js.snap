// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component::TextField Should match snapshoot 1`] = `
<React.Fragment>
  <label
    className="mdc-text-field mdc-text-field--upgraded"
    onClick={[Function]}
    onKeyDown={[Function]}
    style={Object {}}
  >
    <input
      className="mdc-text-field__input"
      disabled={false}
      placeholder={null}
    />
    <FloatingLabel
      cssOnly={false}
      float={false}
      label="test"
      preFilled={false}
      shake={false}
    />
    <LineRipple
      cssOnly={false}
    />
  </label>
</React.Fragment>
`;

exports[`component::TextField Should match snapshoot aria-controls 1`] = `
<React.Fragment>
  <label
    className="mdc-text-field mdc-text-field--upgraded"
    onClick={[Function]}
    onKeyDown={[Function]}
    style={Object {}}
  >
    <input
      aria-controls="test"
      className="mdc-text-field__input"
      disabled={false}
      placeholder={null}
    />
    <FloatingLabel
      cssOnly={false}
      float={false}
      label="test"
      preFilled={false}
      shake={false}
    />
    <LineRipple
      cssOnly={false}
    />
  </label>
</React.Fragment>
`;

exports[`component::TextField Should match snapshoot box 1`] = `
<React.Fragment>
  <label
    className="mdc-text-field mdc-text-field--upgraded mdc-text-field--box"
    onClick={[Function]}
    onKeyDown={[Function]}
    style={Object {}}
  >
    <input
      className="mdc-text-field__input"
      disabled={false}
      placeholder={null}
    />
    <FloatingLabel
      cssOnly={false}
      float={false}
      label="test"
      preFilled={false}
      shake={false}
    />
    <LineRipple
      cssOnly={false}
    />
  </label>
</React.Fragment>
`;

exports[`component::TextField Should match snapshoot cssOnly 1`] = `
<React.Fragment>
  <div
    className="mdc-text-field"
    onClick={[Function]}
    onKeyDown={[Function]}
    style={Object {}}
  >
    <input
      className="mdc-text-field__input"
      disabled={false}
      placeholder="test"
    />
    <FloatingLabel
      cssOnly={true}
      float={false}
      label="test"
      preFilled={false}
      shake={false}
    />
    <LineRipple
      cssOnly={true}
    />
  </div>
</React.Fragment>
`;

exports[`component::TextField Should match snapshoot dense 1`] = `
<React.Fragment>
  <label
    className="mdc-text-field mdc-text-field--upgraded mdc-text-field--dense"
    onClick={[Function]}
    onKeyDown={[Function]}
    style={Object {}}
  >
    <input
      className="mdc-text-field__input"
      disabled={false}
      placeholder={null}
    />
    <FloatingLabel
      cssOnly={false}
      float={false}
      label="test"
      preFilled={false}
      shake={false}
    />
    <LineRipple
      cssOnly={false}
    />
  </label>
</React.Fragment>
`;

exports[`component::TextField Should match snapshoot disabled 1`] = `
<React.Fragment>
  <label
    className="mdc-text-field mdc-text-field--upgraded mdc-text-field--disabled"
    onClick={[Function]}
    onKeyDown={[Function]}
    style={Object {}}
  >
    <input
      className="mdc-text-field__input"
      disabled={true}
      placeholder={null}
    />
    <FloatingLabel
      cssOnly={false}
      float={false}
      label="test"
      preFilled={false}
      shake={false}
    />
    <LineRipple
      cssOnly={false}
    />
  </label>
</React.Fragment>
`;

exports[`component::TextField Should match snapshoot fullWidth 1`] = `
<React.Fragment>
  <label
    className="mdc-text-field mdc-text-field--upgraded mdc-text-field--fullwidth"
    onClick={[Function]}
    onKeyDown={[Function]}
    style={Object {}}
  >
    <input
      className="mdc-text-field__input"
      disabled={false}
      placeholder="test"
    />
    <LineRipple
      cssOnly={false}
    />
  </label>
</React.Fragment>
`;

exports[`component::TextField Should match snapshoot helperText 1`] = `
<React.Fragment>
  <label
    className="mdc-text-field mdc-text-field--upgraded"
    onClick={[Function]}
    onKeyDown={[Function]}
    style={Object {}}
  >
    <input
      aria-controls="test"
      className="mdc-text-field__input"
      disabled={false}
      placeholder={null}
    />
    <FloatingLabel
      cssOnly={false}
      float={false}
      label="test"
      preFilled={false}
      shake={false}
    />
    <LineRipple
      cssOnly={false}
    />
  </label>
  <TextFieldHelperText
    cssOnly={false}
    id="test"
    persistent={false}
    text="this is a helper text"
    validation={false}
  />
</React.Fragment>
`;

exports[`component::TextField Should match snapshoot icon leading 1`] = `
<React.Fragment>
  <label
    className="mdc-text-field mdc-text-field--upgraded mdc-text-field--box mdc-text-field--with-leading-icon"
    onClick={[Function]}
    onKeyDown={[Function]}
    style={Object {}}
  >
    <TextFieldIcon
      className="material-icons"
      clickable={false}
      cssOnly={false}
      disabled={false}
    >
      delete
    </TextFieldIcon>
    <input
      className="mdc-text-field__input"
      disabled={false}
      placeholder={null}
    />
    <FloatingLabel
      cssOnly={false}
      float={false}
      label="test"
      preFilled={false}
      shake={false}
    />
    <LineRipple
      cssOnly={false}
    />
  </label>
</React.Fragment>
`;

exports[`component::TextField Should match snapshoot icon trailing 1`] = `
<React.Fragment>
  <label
    className="mdc-text-field mdc-text-field--upgraded mdc-text-field--outlined mdc-text-field--with-trailing-icon"
    onClick={[Function]}
    onKeyDown={[Function]}
    style={Object {}}
  >
    <input
      className="mdc-text-field__input"
      disabled={false}
      placeholder={null}
    />
    <FloatingLabel
      cssOnly={false}
      float={false}
      label="test"
      preFilled={false}
      shake={false}
    />
    <TextFieldIcon
      className="material-icons"
      clickable={false}
      cssOnly={false}
      disabled={false}
    >
      delete
    </TextFieldIcon>
    <NotchedOutline
      cssOnly={false}
    />
  </label>
</React.Fragment>
`;

exports[`component::TextField Should match snapshoot invalid 1`] = `
<React.Fragment>
  <label
    className="mdc-text-field mdc-text-field--upgraded"
    onClick={[Function]}
    onKeyDown={[Function]}
    style={Object {}}
  >
    <input
      className="mdc-text-field__input"
      disabled={false}
      placeholder={null}
    />
    <FloatingLabel
      cssOnly={false}
      float={false}
      label="test"
      preFilled={false}
      shake={false}
    />
    <LineRipple
      cssOnly={false}
    />
  </label>
</React.Fragment>
`;

exports[`component::TextField Should match snapshoot multiline 1`] = `
<React.Fragment>
  <label
    className="mdc-text-field mdc-text-field--upgraded mdc-text-field--textarea"
    onClick={[Function]}
    onKeyDown={[Function]}
    style={Object {}}
  >
    <textarea
      className="mdc-text-field__input"
      disabled={false}
      placeholder={null}
    />
    <FloatingLabel
      cssOnly={false}
      float={false}
      label="test"
      preFilled={false}
      shake={false}
    />
  </label>
</React.Fragment>
`;

exports[`component::TextField Should match snapshoot outlined 1`] = `
<React.Fragment>
  <label
    className="mdc-text-field mdc-text-field--upgraded mdc-text-field--outlined"
    onClick={[Function]}
    onKeyDown={[Function]}
    style={Object {}}
  >
    <input
      className="mdc-text-field__input"
      disabled={false}
      placeholder={null}
    />
    <FloatingLabel
      cssOnly={false}
      float={false}
      label="test"
      preFilled={false}
      shake={false}
    />
    <NotchedOutline
      cssOnly={false}
    />
  </label>
</React.Fragment>
`;

exports[`component::TextField Should match snapshoot persistentHelperText 1`] = `
<React.Fragment>
  <label
    className="mdc-text-field mdc-text-field--upgraded"
    onClick={[Function]}
    onKeyDown={[Function]}
    style={Object {}}
  >
    <input
      aria-controls="test"
      className="mdc-text-field__input"
      disabled={false}
      placeholder={null}
    />
    <FloatingLabel
      cssOnly={false}
      float={false}
      label="test"
      preFilled={false}
      shake={false}
    />
    <LineRipple
      cssOnly={false}
    />
  </label>
  <TextFieldHelperText
    cssOnly={false}
    id="test"
    persistent={true}
    text="this is a helper text"
    validation={false}
  />
</React.Fragment>
`;

exports[`component::TextField Should match snapshoot validationHelperText 1`] = `
<React.Fragment>
  <label
    className="mdc-text-field mdc-text-field--upgraded"
    onClick={[Function]}
    onKeyDown={[Function]}
    style={Object {}}
  >
    <input
      className="mdc-text-field__input"
      disabled={false}
      placeholder={null}
    />
    <FloatingLabel
      cssOnly={false}
      float={false}
      label="test"
      preFilled={false}
      shake={false}
    />
    <LineRipple
      cssOnly={false}
    />
  </label>
  <TextFieldHelperText
    cssOnly={false}
    persistent={false}
    text="this is a helper text"
    validation={true}
  />
</React.Fragment>
`;

exports[`component::TextField Should match snapshoot value 1`] = `
<React.Fragment>
  <label
    className="mdc-text-field mdc-text-field--upgraded mdc-text-field--upgraded"
    onClick={[Function]}
    onKeyDown={[Function]}
    style={Object {}}
  >
    <input
      className="mdc-text-field__input"
      disabled={false}
      placeholder={null}
      value="test"
    />
    <FloatingLabel
      cssOnly={false}
      float={false}
      label="test"
      preFilled={false}
      shake={false}
    />
    <LineRipple
      cssOnly={false}
    />
  </label>
</React.Fragment>
`;
